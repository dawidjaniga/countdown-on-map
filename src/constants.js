export const WINDOW_ROUTE_POINTS_KEY = '__MapCountdownRoutePoints'
