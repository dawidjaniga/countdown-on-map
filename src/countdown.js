export default class Countdown {
  constructor (containerElement) {
    this.containerElement = containerElement
    this.countdownContainer = document.createElement('div')
    this.countdownContainer.classList.add('map-countdown__countdown')
    this.meta = new Date(2019, 6, 13, 11)
    this.oneDayMilliseconds = 86400000
    this.beginOfYear = new Date(this.meta.getFullYear(), 0, 0)
    this.metaDaysNumberDiff = this.meta - this.beginOfYear
    this.metaDayNumberInYear = Math.floor(
      this.metaDaysNumberDiff / this.oneDayMilliseconds
    )
    this.elements = {
      days: {
        testId: 'map-countdown-days',
        text: 'Dni'
      },
      hours: {
        testId: 'map-countdown-hours',
        text: 'Godzin'
      },
      minutes: {
        testId: 'map-countdown-minutes',
        text: 'Minut'
      },
      seconds: {
        testId: 'map-countdown-seconds',
        text: 'Sekund'
      }
    }

    const fragment = document.createDocumentFragment()

    Object.keys(this.elements).forEach(itemName => {
      const elementFragment = document.createDocumentFragment()
      const item = this.elements[itemName]
      const element = document.createElement('div')
      const number = this.createElement('div', item.testId)
      const text = document.createElement('h4')
      element.classList.add('map-countdown__item')
      text.classList.add('map-countdown__label')
      text.textContent = item.text
      elementFragment.appendChild(element)
      element.appendChild(number)
      element.appendChild(text)

      number.classList.add(
        'map-countdown__number',
        `map-countdown__number--${itemName}`
      )
      fragment.appendChild(elementFragment)
      item.element = element
    })

    this.countdownContainer.appendChild(fragment)
    this.containerElement.appendChild(this.countdownContainer)
    this.events = {}
    this.recountTime()
    this.counterHandler = setInterval(() => this.recountTime(), 1000)
  }

  addEventListener (name, callback) {
    if (!this.events[name]) {
      this.events[name] = []
    }

    this.events[name].push(callback)
  }

  dispatchEvent (name, args) {
    const eventListeners = this.events[name]

    if (eventListeners) {
      eventListeners.forEach(listener => listener(args))
    }
  }

  createElement (tag, testId) {
    const element = document.createElement(tag)
    element.setAttribute('data-testid', testId)
    return element
  }

  setElementValue (name, value) {
    const number = this.elements[name].element.querySelector('div')
    number.textContent = value
  }

  recountTime () {
    var now = new Date()
    var diff = new Date(this.meta - now)
    var daysNumberDiff = new Date(now - this.beginOfYear)
    var currentDayNumberInYear = Math.floor(
      daysNumberDiff / this.oneDayMilliseconds
    )

    var days = this.metaDayNumberInYear - currentDayNumberInYear
    var hours = diff.getHours()
    var minutes = diff.getMinutes()
    var seconds = diff.getSeconds()
    var daysRatio = currentDayNumberInYear / this.metaDayNumberInYear
    var hoursRatio = diff.getHours() / 24
    var minutesRatio = diff.getMinutes() / 60
    var secondsRatio = diff.getSeconds() / 60

    this.dispatchEvent('countdown:recount', {
      days: daysRatio,
      hours: hoursRatio,
      minutes: minutesRatio,
      seconds: secondsRatio
    })

    this.setElementValue('days', days)
    this.setElementValue('hours', hours)
    this.setElementValue('minutes', minutes)
    this.setElementValue('seconds', seconds)
  }
}
