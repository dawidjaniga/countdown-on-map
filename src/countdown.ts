const defaultTranslations = {
  daysOne: 'Day',
  daysMany: 'Days',
  hoursOne: 'Hour',
  hoursMany: 'Hours',
  minutesOne: 'Minute',
  minutesMany: 'Minutes',
  secondsOne: 'Second',
  secondsMany: 'Seconds',
  afterCountdown: 'The event has already took place'
}

interface ICountdownSettings {
  containerElement: HTMLElement,
  meta: string,
  translations: object,
}

interface MapLike<T> {
  [key: string]: T
}

type Element = {
  testId: string,
  element: HTMLElement,
  number: HTMLElement,
  label: HTMLElement
}

type Elements = {
  [index: string]: Element
}

type Event = () => void

type Events = {
  [index: string]: Array<Event>
}

type PluralTranslation = {
  one: string,
  many: string,
}

type Translations = {
  [key: string]: string
}


export default class Countdown {
  private containerElement: HTMLElement
  private translations: Translations
  private countdownContainer: HTMLElement
  private meta: Date
  private oneDayMilliseconds: number
  private beginOfYear: Date
  private metaDaysNumberDiff: number
  private metaDayNumberInYear: number
  private elements: Elements
  private events: Events
  private counterHandler: number

  constructor({ containerElement, meta, translations }: ICountdownSettings) {
    this.containerElement = containerElement
    this.translations = { ...defaultTranslations, ...translations }
    this.elements = {}
    this.countdownContainer = document.createElement('div')
    this.countdownContainer.classList.add('map-countdown__countdown')
    this.meta = new Date(meta)
    this.oneDayMilliseconds = 86400000
    this.beginOfYear = new Date(this.meta.getFullYear(), 0, 0)
    this.metaDaysNumberDiff = this.meta.valueOf() - this.beginOfYear.valueOf()
    this.metaDayNumberInYear = Math.floor(
      this.metaDaysNumberDiff / this.oneDayMilliseconds
    )
    const fragment = document.createDocumentFragment()
    const elements = ['days', 'hours', 'minutes', 'seconds']

    elements.forEach((itemName: string) => {
      const elementFragment = document.createDocumentFragment()
      const item = this.elements[itemName]
      const testId = `map-countdown-${itemName}`
      const element = this.createElement('div', testId)
      const number = document.createElement('div')
      const label = document.createElement('h4')
      element.classList.add('map-countdown__item')
      label.classList.add('map-countdown__label')
      elementFragment.appendChild(element)
      element.appendChild(number)
      element.appendChild(label)

      number.classList.add(
        'map-countdown__number',
        `map-countdown__number--${itemName}`
      )
      fragment.appendChild(elementFragment)
      this.elements[itemName] = {
        element: element,
        number: number,
        label: label,
        testId
      }
    })

    this.countdownContainer.appendChild(fragment)
    this.containerElement.appendChild(this.countdownContainer)
    this.events = {}
    this.recountTime()
    this.counterHandler = window.setInterval(() => this.recountTime(), 1000)
  }

  addEventListener(name: string, callback: () => void) {
    if (!this.events[name]) {
      this.events[name] = []
    }

    this.events[name].push(callback)
  }

  dispatchEvent(name: string, data: object) {
    const eventListeners = this.events[name]

    if (eventListeners) {
      eventListeners.forEach((listener: (data: object) => void) => listener(data))
    }
  }

  createElement(tag: string, testId: string) {
    const element = document.createElement(tag)
    element.setAttribute('data-testid', testId)
    return element
  }

  setElementValue(name: string, value: number) {
    const element = this.elements[name]
    const translationKey = value === 1 ? `${name}One` : `${name}Many`
    const label = this.translations[translationKey]
    element.number.textContent = String(value)
    element.label.textContent = label
  }

  recountTime() {
    const now = new Date()
    const timeLeft = this.meta.valueOf() - now.valueOf()
    const countdownFinished = timeLeft < 0
    const timeLeftDate = new Date(timeLeft)
    const daysNumberDiff = new Date(now.valueOf() - this.beginOfYear.valueOf())
    const currentDayNumberInYear = Math.floor(
      daysNumberDiff.valueOf() / this.oneDayMilliseconds
    )

    if (countdownFinished) {
      this.finishCountdown()
    } else {
      const days = this.metaDayNumberInYear - currentDayNumberInYear
      const hours = timeLeftDate.getHours()
      const minutes = timeLeftDate.getMinutes()
      const seconds = timeLeftDate.getSeconds()
      const daysRatio = currentDayNumberInYear / this.metaDayNumberInYear
      const hoursRatio = timeLeftDate.getHours() / 24
      const minutesRatio = timeLeftDate.getMinutes() / 60
      const secondsRatio = timeLeftDate.getSeconds() / 60

      this.dispatchEvent('countdown:recount', {
        days: daysRatio,
        hours: hoursRatio,
        minutes: minutesRatio,
        seconds: secondsRatio
      })

      this.setElementValue('days', days)
      this.setElementValue('hours', hours)
      this.setElementValue('minutes', minutes)
      this.setElementValue('seconds', seconds)
    }
  }

  finishCountdown() {
    const afterCountdown = this.createElement('h2', 'map-countdown-title')
    afterCountdown.classList.add('map-countdown__title')
    afterCountdown.textContent = this.translations.afterCountdown
    clearInterval(this.counterHandler)
    this.countdownContainer.innerHTML = ''
    this.countdownContainer.appendChild(afterCountdown)
  }
}
