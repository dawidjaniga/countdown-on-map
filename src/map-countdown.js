import Countdown from './countdown'
import Map from './map/map'

export default class MapCountdown {
  constructor ({ selector, routePoints }) {
    this.countdown = new Countdown(selector)
    this.map = new Map({
      key: 'AIzaSyCYkWHZM0ZdO1JeJGBqo44wLlQz31lh-zM',
      selector: '#map'
    })
    this.map.setRoutePoints(routePoints)
    this.countdown.addEventListener(
      'countdown:recount',
      this.updateMap.bind(this)
    )
  }

  updateMap (ratios) {
    this.map.updatePolygons(...Object.values(ratios))
  }
}
