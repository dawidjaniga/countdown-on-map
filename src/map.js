export default class Map {
  constructor () {
    // this.loadMaps()
  }
  updatePolygon (name, value) {
    // console.log('name:', name)
  }

  //   async loadMaps () {
  //     console.log('loadMaps:')
  //     try {
  //       const googleMaps = await loadGoogleMapsApi({
  //         key: 'AIzaSyCYkWHZM0ZdO1JeJGBqo44wLlQz31lh-zM',
  //         libraries: ['drawing']
  //       })
  //       this.map = new googleMaps.Map(document.querySelector('#map'), {
  //         center: {
  //           lat: 40.7484405,
  //           lng: -73.9944191
  //         },
  //         zoom: 12
  //       })
  //     } catch (e) {
  //       console.error('Maps errors', e)
  //     }
  //   }
}
