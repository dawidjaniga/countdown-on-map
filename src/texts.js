export const ROUTE_OPTIONS_MISSING_ERROR = `MapCountdown: route points are missing.
Did you include routePoints.js in head section?
Check for more information: https://github.com/dawidjaniga/map-countdown#add-mapcountdown`
