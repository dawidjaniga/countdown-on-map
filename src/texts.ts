export const ROUTE_OPTIONS_MISSING_ERROR = `MapCountdown: route points are missing.
Did you include routePoints.js in head section?
Check for more information: https://github.com/dawidjaniga/map-countdown#add-mapcountdown`
export const CONTAINER_SELECTOR_ERROR = `MapCountdown: no element found for passed selector
Check element with specified selector exists or MapCountdown is initialized after DOMContentLoaded event.`
