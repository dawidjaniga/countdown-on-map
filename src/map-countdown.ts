import Countdown, { Ratios } from './countdown'
import Map from './map/map'
import { WINDOW_ROUTE_POINTS_KEY } from './constants'
import { ROUTE_OPTIONS_MISSING_ERROR, CONTAINER_SELECTOR_ERROR } from './texts'
import './style.css'

interface MapCountdownSettings {
  selector: string,
  key: string,
  meta: string,
  translations: object,
}

export default class MapCountdown {
  containerElement: HTMLElement | null
  countdown!: Countdown
  map!: Map

  constructor({ selector, key, meta, translations }: MapCountdownSettings) {
    this.containerElement = document.querySelector(selector)

    if (!window[WINDOW_ROUTE_POINTS_KEY]) {
      console.error(ROUTE_OPTIONS_MISSING_ERROR)
      return
    }

    if (!this.containerElement) {
      console.error(CONTAINER_SELECTOR_ERROR)
      return
    }

    this.containerElement.classList.add('map-countdown')
    this.countdown = new Countdown({
      containerElement: this.containerElement,
      meta,
      translations
    })
    this.map = new Map({
      key,
      containerElement: this.containerElement
    })
    this.map.setRoutePoints(window[WINDOW_ROUTE_POINTS_KEY])
    this.attachEvents()

    delete window[WINDOW_ROUTE_POINTS_KEY]
  }
  attachEvents() {
    this.countdown.addEventListener(
      'countdown:recount',
      this.updateMap.bind(this)
    )
  }

  updateMap(ratios: Ratios) {
    this.map.updatePolygons(ratios)
  }
}
