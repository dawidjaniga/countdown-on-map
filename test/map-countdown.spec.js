/* global document, jest, describe, it, expect, beforeEach, afterAll */
import cloneDeep from 'lodash/cloneDeep'
import MapCountdown from '../src/map-countdown'
import routePoints from '../test/__fixtures__/routePoints'
import { WINDOW_ROUTE_POINTS_KEY } from '../src/constants'
import { ROUTE_OPTIONS_MISSING_ERROR } from '../src/texts'
jest.mock('../src/map/map')
jest.mock('../src/countdown')
const originalDocument = cloneDeep(document)

/* eslint-disable no-global-assign */
describe('MapCountdown', () => {
  const spiedConsole = jest.spyOn(console, 'error').mockImplementation()
  const key = 'google-generated-api-key'
  const options = {
    selector: '#countdown',
    key: key,
    meta: '2019-07-13 11:00:00'
  }

  beforeEach(() => {
    document = cloneDeep(originalDocument)
    window[WINDOW_ROUTE_POINTS_KEY] = routePoints
  })

  afterAll(() => {
    document = cloneDeep(originalDocument)
    spiedConsole.mockRestore()
    delete window[WINDOW_ROUTE_POINTS_KEY]
  })

  it('should construct MapCountdown in provided container', () => {
    document.body.innerHTML = `<div id="countdown"></div>`
    new MapCountdown(options) // eslint-disable-line no-new

    expect(document.body).toMatchSnapshot()
  })

  it('should delete routePoints from window', () => {
    document.body.innerHTML = `<div id="countdown"></div>`
    new MapCountdown(options) // eslint-disable-line no-new
    expect(window[WINDOW_ROUTE_POINTS_KEY]).not.toBeDefined()
  })

  it('updateMap() should run updatePolygons on Map', () => {
    document.body.innerHTML = `<div id="countdown"></div>`
    const mapCountdown = new MapCountdown(options)
    const args = {
      days: 1,
      hours: 2
    }
    mapCountdown.updateMap(args)
    expect(mapCountdown.map.updatePolygons).toBeCalledWith(
      ...Object.values(args)
    )
  })

  it('should print error in console when routePoints is missing', () => {
    delete window[WINDOW_ROUTE_POINTS_KEY]
    document.body.innerHTML = `<div id="countdown"></div>`
    new MapCountdown(options) // eslint-disable-line no-new

    expect(spiedConsole).toBeCalledWith(ROUTE_OPTIONS_MISSING_ERROR)
  })
})

/* eslint-enable no-global-assign */
