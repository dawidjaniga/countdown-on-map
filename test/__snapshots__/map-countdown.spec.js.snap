// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapCountdown should construct MapCountdown in provided container 1`] = `
<body>
  <div
    class="map-countdown"
    id="countdown"
  />
</body>
`;
